File: client/py/synnax/segment/iterator.py
from enum import Enum

from freighter import EOF, ExceptionPayload, Payload, Stream, StreamClient

from synnax.channel.registry import ChannelRegistry
from synnax.telem import TimeRange, TimeSpan, TimeStamp

from .encoder import NumpyEncoderDecoder
from .payload import SegmentPayload
from .sugared import NumpySegment

AUTO_SPAN = TimeSpan(-1)


class _Command(int, Enum):
    OPEN = 0
    NEXT = 1
    PREV = 2
    SEEK_FIRST = 3
    SEEK_LAST = 4
    SEEK_LE = 5
    SEEK_GE = 6
    VALID = 7
    ERROR = 8


class _ResponseVariant(int, Enum):
    NONE = 0
    ACK = 1
    DATA = 2


class _Request(Payload):
    command: _Command
    span: TimeSpan | None = None
    range: TimeRange | None = None
    stamp: TimeStamp | None = None
    keys: list[str] | None = None


class _Response(Payload):
    variant: _ResponseVariant
    ack: bool
    command: _Command | None = None
    error: ExceptionPayload | None = None
    segments: list[SegmentPayload] | None = None


class CoreIterator:
    """Used to iterate over a databases telemetry in time-order. It should not be
    instantiated directly, and should instead be instantiated using the segment Client.

    Using an iterator is ideal when querying/processing large ranges of data, but is
    relatively complex and difficult to use. If you're looking to retrieve telemetry
    between two timestamps, see the segment Client read method instead.
    """

    _ENDPOINT = "/segment/iterate"

    client: StreamClient
    stream: Stream[_Request, _Response]
    values: list[SegmentPayload]
    aggregate: bool

    def __init__(self, client: StreamClient, aggregate: bool = False) -> None:
        self.client = client
        self.aggregate = aggregate

    def open(self, keys: list[str], tr: TimeRange):
        """Opens the iterator, configuring it to iterate over the telemetry in the
        channels with the given keys within the provided time range.

        :param keys: The keys of the channels to iterate over.
        :param tr: The time range to iterate over.
        """
        self.stream = self.client.stream(self._ENDPOINT, _Request, _Response)
        self._exec(command=_Command.OPEN, range=tr, keys=keys)
        self.values = []

    def next(self, span: TimeSpan) -> bool:
        """Reads the next time span of telemetry for each channel in the iterator.

        :param span: The span of time to read. A negative span is equivalent to calling
        prev with the absolute value of the span. If AUTO_SPAN is provided, the iterator
        will automatically determine the span of time to read. This is useful for iterating
        quickly over an entire time range.

        :returns: False if a segment satisfying the request can't be found for a
        particular channel or the iterator has accumulated an error.
        """
        return self._exec(command=_Command.NEXT, span=span)

    def prev(self, span: TimeSpan) -> bool:
        """Reads the previous time span of telemetry for each channel in the iterator.

        :param span: The span of time to read. A negative span is equivalent to calling
        next with the absolute value of the span. If AUTO_SPAN is provided, the iterator
        will automatically determine the span of time to read. This is useful for iterating
        quickly over an entire time range.

        :returns: False if a segment satisfying the request can't be found for a particular
        channel or the iterator has accumulated an error.
        """
        return self._exec(command=_Command.NEXT, span=span)

    def seek_first(self) -> bool:
        """Seeks the iterator to the first segment in the time range, but does not read
        it. Also invalidates the iterator. The iterator will not be considered valid
        until a call to first, last, next, prev, prev_span, next_span, or next_range.

        :returns: False if the iterator is not pointing to a valid segment for a particular
        channel or has accumulated an error.
        """
        return self._exec(command=_Command.SEEK_FIRST)

    def seek_last(self) -> bool:
        """Seeks the iterator to the last segment in the time range, but does not read it.
        Also invalidates the iterator. The iterator will not be considered valid
        until a call to first, last, next, prev, prev_span, next_span, or next_range.

        :returns: False if the iterator is not pointing to a valid segment for a particular
        channel or has accumulated an error.
        """
        return self._exec(command=_Command.SEEK_LAST)

    def seek_lt(self, stamp: TimeStamp) -> bool:
        """Seeks the iterator to the first segment whose start is less than or equal to
        the provided timestamp. Also invalidates the iterator. The iterator will not be
        considered valid until a call to first, last, next, prev, prev_span, next_span, or next_range.

        :returns: False if the iterator is not pointing to a valid segment for a particular
        channel or has accumulated an error.
        """
        return self._exec(command=_Command.SEEK_LE, stamp=stamp)

    def seek_ge(self, stamp: TimeStamp) -> bool:
        """Seeks the iterator to the first segment whose start is greater than or equal to
        the provided timestamp. Also invalidates the iterator. The iterator will not be
        considered valid until a call to first, last, next, prev, prev_span, next_span, or next_range.

        :returns: False if the iterator is not pointing to a valid segment for a particular
        channel or has accumulated an error.
        """
        return self._exec(command=_Command.SEEK_GE, stamp=stamp)

    def valid(self) -> bool:
        """Returns true if the iterator value contains a valid segment, and False otherwise.
        valid most commonly returns false when the iterator is exhausted or has accumulated
        an error.
        """
        return self._exec(command=_Command.VALID)

    def close(self):
        """Close closes the iterator. An iterator MUST be closed after use, and this method
        should probably be placed in a 'finally' block. If the iterator is not closed, it make
        leak resources and threads.
        """
        exc = self.stream.close_send()
        if exc is not None:
            raise exc
        pld, exc = self.stream.receive()
        if not isinstance(exc, EOF):
            raise exc

    def _exec(self, **kwargs) -> bool:
        exc = self.stream.send(_Request(**kwargs))
        if exc is not None:
            raise exc
        if not self.aggregate:
            self.values = []
        while True:
            r, exc = self.stream.receive()
            if exc is not None:
                raise exc
            if r.variant == _ResponseVariant.ACK:
                return r.ack
            self.values += r.segments


class NumpyIterator(CoreIterator):
    """Used to iterate over a databases telemetry in time-order. It should not be
    instantiated directly, and should instead be instantiated using the segment Client.

    Using an iterator is ideal when querying/processing large ranges of data, but is
    relatively complex and difficult to use. If you're looking to retrieve telemetry
    between two timestamps, see the segment Client read method instead.
    """

    decoder: NumpyEncoderDecoder
    channels: ChannelRegistry

    def __init__(
        self,
        transport: StreamClient,
        channels: ChannelRegistry,
        aggregate: bool = False,
    ):
        super().__init__(transport, aggregate)
        self.decoder = NumpyEncoderDecoder()
        self.channels = channels

    def open(self, keys: list[str], tr: TimeRange) -> None:
        super().open(keys, tr)

    @property
    def value(self) -> dict[str, NumpySegment]:
        """
        :returns: The current iterator value as a dictionary whose keys are channels
        and values are segments containing telemetry at the current iterator position.
        """
        decoded = []
        self.values.sort(key=lambda v: v.start)
        res = dict()
        for i, seg in enumerate(self.values):
            decoded.append(self.decoder.decode(self.channels.get(seg.channel_key), seg))
        for i, dec in enumerate(decoded):
            if dec.channel.key in res:
                res[dec.channel.key].extend(dec)
            else:
                res[dec.channel.key] = dec
        return res

File: client/ts/src/lib/segment/iterator.ts
import {
  EOF,
  ErrorPayloadSchema,
  Stream,
  StreamClient,
} from '@synnaxlabs/freighter';
import { z } from 'zod';

import { ChannelPayload } from '../channel/payload';
import Registry from '../channel/registry';
import { TimeRange } from '../telem';

import { SegmentPayload, SegmentPayloadSchema } from './payload';
import TypedSegment from './typed';

enum Command {
  Open = 0,
  Next = 1,
  Prev = 2,
  First = 3,
  Last = 4,
  NextSpan = 5,
  PrevSpan = 6,
  NextRange = 7,
  Valid = 8,
  Error = 9,
  SeekFirst = 10,
  SeekLast = 11,
  SeekLT = 12,
  SeekGE = 13,
}

enum ResponseVariant {
  None = 0,
  Ack = 1,
  Data = 2,
}

const RequestSchema = z.object({
  command: z.nativeEnum(Command),
  span: z.number().optional(),
  range: z.instanceof(TimeRange).optional(),
  stamp: z.number().optional(),
  keys: z.string().array().optional(),
});

type Request = z.infer<typeof RequestSchema>;

const ResponseSchema = z.object({
  variant: z.nativeEnum(ResponseVariant),
  ack: z.boolean(),
  command: z.nativeEnum(Command),
  error: ErrorPayloadSchema.optional(),
  segments: SegmentPayloadSchema.array().nullable(),
});

type Response = z.infer<typeof ResponseSchema>;

/**
 * Used to iterate over a clusters telemetry in time-order. It should not be
 * instantiated directly, and should instead be instantiated via the SegmentClient.
 *
 * Using an iterator is ideal when querying/processing large ranges of data, but
 * is relatively complex and difficult to use. If you're looking to retrieve
 *  telemetry between two timestamps, see the SegmentClient.read method.
 */
export class CoreIterator {
  private static ENDPOINT = '/segment/iterate';
  private client: StreamClient;
  private stream: Stream<Request, Response> | undefined;
  private readonly aggregate: boolean = false;
  values: SegmentPayload[] = [];

  constructor(client: StreamClient, aggregate = false) {
    this.client = client;
    this.aggregate = aggregate;
  }

  /**
   * Opens the iterator, configuring it to iterate over the telemetry in the
   * channels with the given keys within the provided time range.
   *
   * @param tr - The time range to iterate over.
   * @param keys - The keys of the channels to iterate over.
   */
  async open(tr: TimeRange, keys: string[]) {
    this.stream = await this.client.stream(
      CoreIterator.ENDPOINT,
      RequestSchema,
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      ResponseSchema
    );
    await this.execute({ command: Command.Open, keys, range: tr });
    this.values = [];
  }

  /**
   * Reads the next segment for each channel in the iterator.
   *
   * @returns false if the next segment can't be found for one or more channels or
   * the iterator has accumulated an error.
   */
  async next(): Promise<boolean> {
    return this.execute({ command: Command.Next });
  }

  /**
   * Reads the previous segment for each channel in the iterator.
   *
   * @returns false if the next segment can't be found for one or more channels or
   * the iterator has accumulated an error.
   */
  async prev(): Promise<boolean> {
    return this.execute({ command: Command.Prev });
  }

  /**
   * Seeks to the beginning of the time range and reads the first segment of each
   * channel in the iterator.
   *
   * @returns false if no segments exists in the time range for a particular channel
   * or the iterator has accumulated an error.
   */
  async first(): Promise<boolean> {
    return this.execute({ command: Command.First });
  }

  /**
   * Seeks to the end of the time range and reads the last segment of each channel
   * in the iterator.
   *
   * @returns false if no segments exists in the time range for a particular channel,
   * or the iterator has accumulated an error.
   */
  async last(): Promise<boolean> {
    return this.execute({ command: Command.Last });
  }

  /**
   * Reads the next time span of telemetry for each channel in the iterator.
   *
   * @returns false if a segment satisfying the request can't be found for a
   * particular channel or the iterator has accumulated an error.
   */
  async nextSpan(span: number): Promise<boolean> {
    return this.execute({ command: Command.NextSpan, span });
  }

  /**
   * Reads the previous time span of telemetry for each channel in the iterator.
   *
   * @returns false if a segment satisfying the request can't be found for a particular
   * channel or the iterator has accumulated an error.
   */
  async prevSpan(span: number): Promise<boolean> {
    return this.execute({ command: Command.PrevSpan, span });
  }

  /**
   * Seeks the iterator to the start of the time range and reads the telemetry within
   * the range for each channel.
   *
   * @returns: False if a segment satisfying the request can't be found for a particular
   * channel or the iterator has accumulated an error.
   */
  async nextRange(range: TimeRange): Promise<boolean> {
    return this.execute({ command: Command.NextRange, range });
  }

  /**
   * Seeks the iterator to the first segment in the time range, but does not read
   * it. Also invalidates the iterator. The iterator will not be considered valid
   * until a call to first, last, next, prev, prev_span, next_span, or next_range.
   *
   * @returns false if the iterator is not pointing to a valid segment for a particular
   * channel or has accumulated an error.
   */
  async seekFirst(): Promise<boolean> {
    return this.execute({ command: Command.SeekFirst });
  }

  /** Seeks the iterator to the last segment in the time range, but does not read it.
   * Also invalidates the iterator. The iterator will not be considered valid
   * until a call to first, last, next, prev, prev_span, next_span, or next_range.
   *
   * @returns false if the iterator is not pointing to a valid segment for a particular
   * channel or has accumulated an error.
   */
  async seekLast(): Promise<boolean> {
    return this.execute({ command: Command.SeekLast });
  }

  /**
   * Seeks the iterator to the first segment whose start is less than or equal to
   * the provided timestamp. Also invalidates the iterator. The iterator will not be
   * considered valid until a call to first, last, next, prev, prev_span, next_span, or next_range.
   *
   * @returns false if the iterator is not pointing to a valid segment for a particular
   * channel or has accumulated an error.
   */
  async seekLT(stamp: number): Promise<boolean> {
    return this.execute({ command: Command.SeekLT, stamp });
  }

  /**
   * Seeks the iterator to the first segment whose start is greater than or equal to
   * the provided timestamp. Also invalidates the iterator. The iterator will not be
   * considered valid until a call to first, last, next, prev, prev_span, next_span, or next_range.
   *
   * @returns false if the iterator is not pointing to a valid segment for a particular
   * channel or has accumulated an error.
   */
  async seekGE(stamp: number): Promise<boolean> {
    return this.execute({ command: Command.SeekGE, stamp });
  }

  /**
   * @returns true if the iterator value contains a valid segment, and fale otherwise.
   * valid most commonly returns false when the iterator is exhausted or has
   * accumulated an error.
   */
  async valid(): Promise<boolean> {
    return this.execute({ command: Command.Valid });
  }

  /**
   * Closes the iterator. An iterator MUST be closed after use, and this method
   * should probably be placed in a 'finally' block. If the iterator is not closed,
   * it may leak resources.
   */
  async close() {
    if (!this.stream)
      throw new Error('iterator.open() must be called before any other method');
    this.stream.closeSend();
    const [, exc] = await this.stream.receive();
    if (!(exc instanceof EOF)) throw exc;
  }

  private async execute(request: Request): Promise<boolean> {
    if (!this.stream)
      throw new Error('iterator.open() must be called before any other method');
    const err = this.stream.send(request);
    if (err) throw err;
    if (!this.aggregate) this.values = [];
    for (;;) {
      const [res, err] = await this.stream.receive();
      if (err || !res) throw err;
      if (res.variant == ResponseVariant.Ack) return res.ack;
      if (res.segments) this.values.push(...res.segments);
    }
  }
}

export class TypedIterator extends CoreIterator {
  channels: Registry;

  constructor(client: StreamClient, channels: Registry, aggregate = false) {
    super(client, aggregate);
    this.channels = channels;
  }

  async value(): Promise<Record<string, TypedSegment>> {
    const result: Record<string, TypedSegment> = {};
    this.values.sort((a, b) => a.start.valueOf() - b.start.valueOf());
    const keys = this.values.map((v) => v.channelKey);
    const channels = await this.channels.getN(...keys);
    this.values.forEach((v) => {
      const sugared = new TypedSegment(
        channels.find((c) => c.key == v.channelKey) as ChannelPayload,
        v
      );
      if (v.channelKey in result) {
        result[v.channelKey].extend(sugared);
      } else {
        result[v.channelKey] = sugared;
      }
    });
    return result;
  }
}
